jni/src/Foreign/JNI.hs

```diff
@@ -435,11 +459,14 @@ setStatic/**/name/**/Field klass field x = \
     throwIfException env $ \
     [CU.block| void { \
       (*$(JNIEnv *env))->SetStatic/**/name/**/Field($(JNIEnv *env), \
-                                                    $(jclass klass), \
+                                                    $fptr-ptr:(jclass klass), \
                                                     $(jfieldID field), \
                                                     $(c_fieldtype x)); } |]
 
-SET_STATIC_FIELD(Object, JObject, jobject)
+setStaticObjectField :: JClass -> JFieldID -> JObject -> IO ()
+setStaticObjectField x y z =
+    let SET_STATIC_FIELD(Object, (Ptr JObject), jobject)
+    in withForeignPtr (coerce z) (setStaticObjectField x y)
```

Comment:
I'm just reusing the existing idiom that is already used in the rest of the file for the `Boolean` type. If we don't like that idiom we can change it for the whole file in a separate PR (but I doubt the extra clutter - double the macros definitions etc - will have a positive impact on balance).